redact/index.ts

```diff
@@ -0,0 +1,213 @@
+    return { tag: "email" as const };
+  }
+
+  if (entity === "phone-number") {
+    return { tag: "phone-number" as const };
+  }
+
+  if (entity === "ip-address") {
+    return { tag: "ip-address" as const };
+  }
+
+  if (entity === "credit-card-number") {
+    return { tag: "credit-card-number" as const };
+  }
+
+  return {
+    tag: "custom" as const,
+    val: entity,
+  };
+}
+
+function wasmEntitiesToString(entity: SensitiveInfoEntity) {
+  if (entity.tag === "email") {
+    return "email";
+  }
+
+  if (entity.tag === "ip-address") {
+    return "ip-address";
+  }
+
+  if (entity.tag === "credit-card-number") {
+    return "credit-card-number";
+  }
+
+  if (entity.tag === "phone-number") {
+    return "phone-number";
+  }
+
+  return entity.val;
+}
+
+function performReplacementInText(
+  text: string,
+  replacement: string,
+  start: number,
+  end: number,
+): string {
+  return text.substring(0, start) + replacement + text.substring(end);
+}
+
+function noOpDetect(_tokens: string[]): Array<SensitiveInfoEntity | undefined> {
+  return [];
+}
+function noOpReplace(_input: SensitiveInfoEntity): string | undefined {
+  return undefined;
+}
+
+interface RedactedSensitiveInfoEntity {
+  original: string;
+  redacted: string;
+  start: number;
+  end: number;
+  identifiedType: string;
+}
+
+async function callRedactWasm<
+  const Detect extends DetectSensitiveInfoEntities<CustomEntities> | undefined,
+  const CustomEntities extends string,
+>(
+  candidate: string,
+  options?: RedactOptions<Detect>,
+): Promise<RedactedSensitiveInfoEntity[]> {
+  let convertedDetect = noOpDetect;
+  if (typeof options?.detect === "function") {
+    const detect = options.detect;
+    convertedDetect = (tokens: string[]) => {
+      return detect(tokens)
+        .filter((e) => typeof e !== "undefined")
+        .map((e) => userEntitiesToWasm(e));
+    };
+  }
+
+  let convertedReplace = noOpReplace;
+  if (typeof options?.replace === "function") {
+    const replace = options.replace;
+    convertedReplace = (identifiedType: SensitiveInfoEntity) => {
+      return replace(
+        // @ts-ignore because we know this is coming from Wasm
+        wasmEntitiesToString(identifiedType),
+      );
+    };
+  }
+
+  const wasm = await initializeWasm(convertedDetect, convertedReplace);
+
+  if (typeof wasm !== "undefined") {
+    const skipCustomDetect = typeof options?.detect !== "function";
+    const skipCustomRedact = typeof options?.replace !== "function";
+
+    if (
+      typeof options?.entities !== "undefined" &&
+      !Array.isArray(options?.entities)
+    ) {
+      throw new Error("entities must be an array");
+    } else if (options?.entities?.length === 0) {
+      throw new Error("no entities configured for redaction");
+    }
+
+    const config = {
+      entities: options?.entities?.map(userEntitiesToWasm),
+      contextWindowSize: options?.contextWindowSize,
+      skipCustomDetect,
+      skipCustomRedact,
+    };
```

Comment:
All of this is super sloppy. Let's move it into a function. I wrote this:

```ts
function getWasmOptions<
  const Detect extends DetectSensitiveInfoEntities<CustomEntities> | undefined,
  const CustomEntities extends string,
>(options?: RedactOptions<Detect>) {
  if (typeof options === "object" && options !== null) {
    if (Array.isArray(options.entities)) {
      if (options.entities.length < 1) {
        throw new Error("no entities configured for redaction");
      }
    } else {
      throw new Error("entities must be an array");
    }

    return {
      entities: options.entities.map(userEntitiesToWasm),
      contextWindowSize: options.contextWindowSize || 1,
      skipCustomDetect: typeof options.detect !== "function",
      skipCustomRedact: typeof options.replace !== "function",
    };
  } else {
    return {
      entities: undefined,
      contextWindowSize: 1,
      skipCustomDetect: true,
      skipCustomRedact: true,
    };
  }
}
```

Reply:
Yeah good suggestion, this is a lot cleaner!

Reply:
Good catch on the undefined